text/3729-sized-hierarchy.md

```diff
@@ -0,0 +1,1412 @@
+  [rust#46108][pr_dynsized_rebase],  [rfcs#2984][rfc_pointee_dynsized] and
+  [eRFC: Minimal Custom DSTs via Extern Type (DynSized)][erfc_minimal_custom_dsts_via_extern_type],
+  none of the traits proposed in this RFC are default bounds and therefore do not
+  need to support being relaxed bounds (i.e. no `?ValueSized`), which avoids
+  additional language complexity and backwards compatibility hazards related to
+  relaxed bounds and associated types.
+- In contrast to [rfcs#1524][rfc_custom_dst], [rfc#1993][rfc_opaque_data_structs],
+  [Pre-eRFC: Let's fix DSTs][pre_erfc_fix_dsts], [Pre-RFC: Custom DSTs][prerfc_custom_dst]
+  and [eRFC: Minimal Custom DSTs via Extern Type (DynSized)][erfc_minimal_custom_dsts_via_extern_type],
+  `ValueSized` does not have `size_of_val`/`align_of_val` methods to support 
+  custom DSTs as this would add to the complexity of this proposal and custom DSTs
+  are not this RFC's focus, see the [Custom DSTs][custom-dsts] section later.
+
+## Why have `Pointee`?
+[why-have-pointee]: #why-have-pointee
+
+It may seem that re-using the `Pointee` trait at the bottom of the trait hierarchy
+is unnecessary as this is equivalent to the absense of any bounds whatsoever, but
+having an `Pointee` trait is necessary to enable the meaning of `?Sized` to be re-defined
+to be equivalent to `const ValueSized` and avoid complicated behaviour change over an edition.
+
+Without `Pointee`, if a user wanted to remove all sizedness bounds from a generic
+parameter then they would have two options:
+
+1. Introduce new relaxed bounds (i.e. `?ValueSized`), which has been found
+   unacceptable in previous RFCs ([rfcs#2255][issue_more_implicit_bounds]
+   summarizes these discussions)
+2. Keep `?Sized`'s existing meaning of removing the implicit `Sized` bound
+      
+   This is the only viable option, but this would complicate changing
+   `size_of_val`'s existing `?Sized` bound:
+
+   Without `Pointee`, `?Sized` would be equivalent to `const ValueSized` until
+   extern types are stabilised (e.g. a `?Sized` bound would accept exactly the
+   same types as a `const ValueSized` bound, but after extern types are introduced,
+   `?Sized` bounds would accept extern types and `const ValueSized` bounds would not).
+   extern types would need to be introduced over an edition and all existing `?Sized`
+   bounds rewritten to `?Sized + const ValueSized`. This is the same mechanism described
+   in [rfcs#3396][rfc_extern_types_v2] to introduce its `MetaSized` trait.
+
+## Why use const traits?
+[why-use-const-traits]: #why-use-const-traits
+
+Previous iterations of this RFC had both linear[^6] and non-linear[^7] trait hierarchies
+which included a `RuntimeSized` trait and did not use const traits. However, both of
+these were found to be backwards-incompatible due to being unable to relax the
+supertrait of `Clone`. Without const traits, it is not possible to represent
```

Comment:
This is mentioned in the top-level RFC description:

> This RFC relies on experimental, yet-to-be-RFC'd const traits, so this is blocked on that. I haven't squashed any of the previous revisions but can do so if/when this is approved.

`const` traits have a *huge* design and implementation space, and it seems like stabilization of such a feature is quite a ways off.

Have you considered whether there's a way, possibly via trait aliases or similar, to create a stable interface for the new `Sized` hierarchy without first stabilizing the `const` trait feature?

Reply:
A sufficiently minimal subset of the `const` traits implementation space that fully enables this proposal may actually be possible to drive to stabilization in 2025, such that I do not think this RFC should be hung on this concern.

I defer to @fee1-dead, @oli-obk, and @compiler-errors on this, of course.

Reply:
Also, trait aliases aren't stable either.

Reply:
From speaking to @compiler-errors, I'm under the impression that the types team have made a lot of progress with the implementation of const traits and feel quite confident in the implementation and that they understand the feature and its implications. I'm sure there's still lots to do but it sounded promising.

I don't think it's possible to implement this without const traits - we considered introducing another trait in the hierarchy `RuntimeSized` in earlier revisions, but it isn't possible to relax the supertrait of `Clone` from `Sized` to `RuntimeSized`, so you need to keep it as `Sized` and rely on something like const traits.

Reply:
To clarify, I'm wondering if it's possible to rely on const traits in a way that doesn't expose to end-users that they're using const traits. This would potentially allow stabilization of the feature before broad stabilization of a (much larger) const traits feature.

Reply:
why not have `const Sized` under a separate feature gate from general const traits and stabilize just `const Sized` first?

Reply:
because we probably also want `~const Sized` which is officially placeholder syntax

Reply:
why not just say `~const Sized` can wait till later, since `const Sized` is needed to express what `Sized` does now, but `~const Sized` is new?

Reply:
The syntax part may be separatable, but the background work (ensuring everything is consistent and forward compatible) required is the same

Reply:
Even if you had only `const Sized`, you would probably want to be able to write functions with `~const Sized` bounds so that when invoked out of a const context, they can accept runtime-sized types.